src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+                output_list.append(Literal(f"{i}", INTEGER_TYPE))
+            return output_list
+
+        # If ref is a reference then we generate the first two values of
+        # the sequence only.
+        output_list.append(start.copy())
+        second_val = BinaryOperation.create(
+                BinaryOperation.Operator.ADD,
+                start.copy(),
+                step.copy())
+        output_list.append(second_val)
+        return output_list
+
+    def _compare_ref_binop(self, ref1, ref2, task1, task2):
+        '''
+        Compares two Reference/BinaryOperations Nodes to check they are a valid
+        dependency structure.
+
+        :param ref1: The first Node to compare.
+        :type ref1: Union[:py:class:`psyclone.psyir.nodes.Reference, \
+                    :py:class:`psyclone.psyir.nodes.BinaryOperation]
+        :param ref2: The second Node to compare.
+        :type ref2: Union[:py:class:`psyclone.psyir.nodes.Reference, \
+                    :py:class:`psyclone.psyir.nodes.BinaryOperation]
+        :param task1: The task containing ref1 as a child.
+        :type task1: :py:class:`psyclone.psyir.nodes.OMPTaskDirective
+        :param task2: The task containing ref2 as a child.
+        :type task2: :py:class:`psyclone.psyir.nodes.OMPTaskDirective
+
+        :raises GenerationError: If ref1 and ref2 are dependencies on the \
+                                 same array, but one does not contain a \
+                                 Reference but the other does.
+        :raises GenerationError: If ref1 and ref2 are dependencies on the \
+                                 same array but are References to different \
+                                 variables.
+        :raises GenerationError: If ref1 and ref2 are dependencies on the \
+                                 same array, but the computed index values \
+                                 are not dependent according to OpenMP.
+
+        :returns: If these two nodes can be used as a valid dependency pair \
+                  in OpenMP
+        :rtype: bool
+        '''
+        # In this case we have two Reference/BinaryOperation as indices.
+        # We need to attempt to find their value set and check the value
+        # set matches.
+        # Find all the nodes before these tasks
+        preceding_t1 = task1.preceding(reverse=True)
+        preceding_t2 = task2.preceding(reverse=True)
+        # Get access list for each ref
+        try:
+            ref1_accesses = self._compute_accesses(ref1, preceding_t1, task1)
+            ref2_accesses = self._compute_accesses(ref2, preceding_t2, task2)
+        except GenerationError:
+            # If we get a Generation error from compute_accesses, then we found
+            # an access that isn't able to be handled by PSyclone, so
+            # dependencies based on it need to be handled by a taskwait
+            return False
+
+        # If the first access in each accesses contains a Reference
```

Comment:
We know from `_compute_accesses` that all References contained in the output are to the same symbol, and if any member of the result is a Reference, the first member has to be.

I can add a comment here to this if you think it would be helpful.